### DISCUSSION TITLE
Leaked Semaphores

### DISCUSSION THREAD

**Discussion by ChipsAHoai** [zauberzeug/nicegui#4129] [Q&A]
### Question

I am having leaked semaphore issues when running nicegui with fastapi.  Initially, I thought it was my code, however, I just tested this on Zauberzeug's repo running the start.sh bash script under examples/fastapi and immediately closed down the uvicorn server and it still shows leaked semaphores.  Is there a solution for this?

**Comment by rodja** [comment]
You are right. I can reproduce the error:

```
cd examples/fastapi
./start.sh dev
Starting Uvicorn server in development mode...
INFO:     Will watch for changes in these directories: ['nicegui/examples/fastapi']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [54107] using WatchFiles
INFO:     Started server process [54109]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [54109]
INFO:     Stopping reloader process [54107]
/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown
```

`^CINFO:     Shutting down` is the line where I press CTRL-C to abort the execution.

**Comment by rodja** [comment]
This is definitely a bug. I created issue https://github.com/zauberzeug/nicegui/issues/4131. Let's continue there.

**Reply by MaxRev-Dev** [reply]
@rodja Hi. I still get this warning.
OS: MacOS Sequoia 15.2
Fresh install into venv. Latest NiceGUI version (2.11.1). Tried a Lightbox example (image gallery). 

```
<path>/.pyenv/versions/3.9.19/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 33 leaked semaphore objects to clean up at shutdown
```

**Reply by rodja** [reply]
Wooo. There was a revert of the fix, but I do not remember why. Let's continue on the issue https://github.com/zauberzeug/nicegui/issues/4131

**Reply by falkoschindler** [reply]
The fix has been reverted because it caused a new problem: #4244.

**Reply by rodja** [reply]
There is a much more elegant fix in #4641 (see c516d03) which hopefully gets merged in the next release.
